matching/README.md

```diff
@@ -0,0 +1,991 @@
+being specialized generally has multiple columns, corresponding to multiple
+inputs to the pattern matching function. We thus must define how those inputs
+are created from the previous term. Further, if any of the constructors
+matched statically by the pattern match compiler have variables that are bound
+as part of the `specializeBy` function, these variables will exist in the
+substitution for that row in the matrix, but they will not exist in the
+substitution generated by the `evaluate` function.
+
+Thus, in order to fully define this optimization, we must address all three
+concerns: how to generate the different inputs to the `evaluate` function, how
+to compile the matrix to a decision tree which generates all the relevant
+bindings for statically-matched terms, and how the `specializeBy` function
+operates.
+
+We begin with the `evaluate` function.
+
+Unlike before, where evaluation took as input a term and a decision tree,
+it now takes a decision tree and an initial substitution. The initial
+substitution contains one binding per entry in the residual map. Bear in mind
+the keys in the residual map are occurrences and the values are terms from the
+right-hand side of a rule, which may contain variables bound by the left-hand
+side of that rule, and may also contain things like functions and such which
+are not traditionally thought of as patterns.
+
+We can construct the initial substitution of the `evaluate` function quite
+easily, as it turns out. Simply substitute the variables from the left-hand
+side of the previous rule into the residuals that were part of that rule's
+right-hand side, evaluating the result according to the same semantics used
+for the right-hand side of rules in your language. Then bind them to the same
+occurrences they were bound to in the residual map itself, and pass that map to
+the `evaluate` function. As before, we start with an empty stack, and the
+algorithm proceeds unchanged from that point.
+
+However, we still need to address the issue of bindings for variables that
+were bound as part of the execution of the `specializeBy` function. It turns
+out the solution is to add a new field to each row in the matrix recording
+the variables that were bound by the `specializeBy` function. Then, in the
+`getLeaf` function, we modify the return value `dt` to return something like
+`MakePattern(o, s, p, dt)` where `o` and `s` are the occurrence and sort
+for the variable that was bound, and `p` is the pattern that it was bound to.
+
+Finally, we must define the `specializeBy` function. We define it quite
+simply in pseudo-code:
+
+```
+def specializeBy(Matrix m, [Term] rhs): (Matrix, [Term])
+  if m has no rows:
+    return (m, rhs)
+  if every term t in rhs is a wildcard pattern or is not a pattern:
+    return (m, rhs)
+  pick the best column j of m such that rhs[j] is a non-wildcard pattern
+  val c = the constructor of rhs[j]
+  return specializeBy(specialize(m, j, c), specialize(rhs, j, c))
```

Comment:
So the iteration process will keep going until it reaches an empty matrix (terminal pattern?) or the RHS is completely wildcard patterns.

Did we do any experiments on this process? Is it ever worth applying the optimisation partially to trade off between code size and runtime performance?

Reply:
I mean, you wouldn't really get much of a space savings by only doing it partially, because you would still have the problem that you have a bunch of different decision trees, one for each rule, that need to be turned into functions.

But we do have some heuristics that say, don't bother generating a decision tree for a particular rule if it isn't much better than the original decision tree. I will add some text here to explain how we do that.